File: RentalSystem.ts
import { RentalItem } from "./RentalItem";
import { Reservation } from "./Reservation";

export class RentalSystem {
  items: RentalItem[] = [];

  addItem(item: RentalItem) {
    this.items.push(item);
  }

  getItems(): RentalItem[] {
    return this.items;
  }

  reserveItem(itemName: string, startDate: Date, days: number): Reservation {
    const item = this.items.find(i => i.name.toLowerCase() === itemName.toLowerCase());

    if (!item) {
      throw new Error("Objet introuvable.");
    }

    if (days < 1) {
      throw new Error("Durée minimale : 1 jour.");
    }

    if (!item.isAvailable(startDate, days)) {
      throw new Error("Objet déjà réservé pendant cette période.");
    }

    const reservation = new Reservation(item, startDate, days);
    item.reservations.push(reservation);
    return reservation;
  }
}

File: RentalSystem.java
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;

public class RentalSystem {
    private List<RentalItem> items;

    public RentalSystem() {
        items = new ArrayList<>();
    }

    public void ajouterObjet(RentalItem item) {
        items.add(item);
    }

    public List<RentalItem> getItems() {
        return items;
    }

    public Reservation reserverObjet(String nomObjet, LocalDate dateDebut, int jours) throws ReservationException {
        for (RentalItem item : items) {
            if (item.getNom().equalsIgnoreCase(nomObjet)) {
                if (jours < 1) {
                    throw new ReservationException("La durée minimale de réservation est de 1 jour.");
                }
                if (item.estDisponible(dateDebut, jours)) {
                    Reservation reservation = new Reservation(item, dateDebut, jours);
                    item.getReservations().add(reservation);
                    return reservation;
                } else {
                    throw new ReservationException(nomObjet + " est déjà réservé pendant cette période.");
                }
            }
        }
        throw new ReservationException("Objet non trouvé.");
    }
}
